keps/sig-api-machinery/3937-declarative-validation/README.md

```diff
@@ -0,0 +1,1077 @@
+
+- It will make it easier to develop, maintain and review APIs.
+- It will make it easier programatically inspect and analyze the API, enabling
+  new tools and improved documentation.
+- It will enable improvements to the API machinery. For example, a feature like
+  ratcheting validation will become more tractable to implement because the
+  feature can be implemented once in the declarative validation subsystem
+  rather than piecemeal across the 15k lines of hand written validation code.
+
+Declarative validation will also benefit Kubernetes users:
+
+- It will give users direct access to the actual API validation rules, which
+  are currently only available to developers willing and able to find and read
+  the hand written validation rules.
+- It will enable clients to perform validation of native types earlier in
+  development worflows ("shift-left"), such as at with a git pre-submit linter.
+- It will improve API composition. In particular CRDs that embed native types
+  (such as PodTemplate), which gain validation of the native type automatically.
+  This has the potential to simplify controller development and improve end
+  user experiences when using CRDs.
+
+Please feel free to try out the
+[prototype](https://github.com/jpbetz/kubernetes/blob/cel-for-native/staging/src/k8s.io/sample-apiserver/pkg/apis/wardle/validation/README.md)
+to get hands on experience with this proposed enhancement.
+
+### Goals
+
+- Vast majority (95%+) of hand written validation are replaced with declarative
+validation. The remaining hand written rules will primarily be rules that _should
+not_ be published as part of the API, usually because server side state is
+involved in the validation decision.
+
+- `types.go` files become the de-facto IDL of Kubernetes for native types.
+It is worthing noting that `+enum` support, `+default` support and similar
+enhancements all moved our API development forward in this direction. This
+enhancement is an attempt to finish that story arc.
+
+- CRDs and native types are validated and published in OpenAPI in a consistent
+and uniform way. Improvement we make to CEL to enable declarative validation of
+native types will be made available to CRD authors and vis-versa.
+
+### Non-Goals
+
+- It is not a goal for _all_ validation rules to be published to OpenAPI. There
+  exist complex, stateful rules that should not be published in OpenAPIv3. These
+  will remain "server side only" validation rules.
+- It is not a goal to require that _all_ validation rules be written in CEL. We
+  will continue to support hand written validation rules indefinitely for use
+  cases that are a poor fit using the existing validation mechanism (or
+  something very similar).
+
+
+## Proposal
+
+Declarative validation will be performed against versioned APIs. This differs from the hand
+written validation, which is evaluated against internal types.
+
+Go IDL tags will be added to support the following declarative validation rules:
```

Comment:
is the list exhaustive?

Reply:
That is the intent, yes. Let me know if I missed anything.